### PR TITLE
SRCH-58

### PR DESCRIPTION
**Pull request by scott-lewis-nsidc**
Updating the code to use the `environments.yaml` file for the configuration (which can still be overridden by environment variables, useful for testing and local work).

### CODE REVIEW

**Review by juliacollins — CHANGES_REQUESTED** [review]

File: lib/search_solr_tools/config/environments.yaml [hunk 1]
```diff
@@ -9,17 +9,26 @@
...
 
+  # Log details.  Can be overridden by environment-specific values
+  :log_file: /var/log/search-solr-tools.log
+  :log_file_level: warn
+  :log_stdout_level: info
+
 :local:
   :host: localhost
   :nsidc_dataset_metadata_url: http://integration.nsidc.org/api/dataset/metadata/
+  :log_file: none
```
[the remark this answers is not in the document]
Comment (scott-lewis-nsidc) [reply lib/search_solr_tools/config/environments.yaml:20]: Ok, I can make that change.

File: README.md [hunk 2]
```diff
@@ -208,6 +212,50 @@ which can be modified, or additional environments can be added by just adding a
 new YAML stanza with the right keys; this new environment can then be used with
 the `--environment` flag when running `search_solr_tools harvest`.
 
+#### Logging
+
+By default, when running the harvest, harvest logs are written to the file
+`/var/log/search-solr-tools.log` (set to `warn` level), as well as to the console
```
[the remark this answers is not in the document]
Comment (scott-lewis-nsidc) [reply README.md:218]: Both MDI and DCS write their logs to /var/logs, which then get rotated by `logrotate` to /share/logs/<projectname>.

CMR Mediator writes to /share/logs directly, though, 

I'll ask Daniel and see if he has a preference.  I was using the MDI/DCS approach for this one.

File: lib/search_solr_tools/logging/sst_logger.rb [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+  def log_environment(new_env)
+    SSTLogger.log_environment(new_env)
+  end
+
+  class << self
+    def logger
+      @logger ||= new_logger
+    end
+
+    def log_environment(new_env)
```
[the remark this answers is not in the document]
Comment (scott-lewis-nsidc) [reply lib/search_solr_tools/logging/sst_logger.rb:22]: I was trying to avoid puts statements, as then you have to add manual flags to each statement.  By using the logger to stdout, you can just do "logger.info" and it will potentially write to one or both places automatically.  No separate checking needed.

That said, I could have a verbose flag set up the stdout log, as opposed to having it be a configuration/environment variable, but that would mean having to pass that verbosity into the logger initialization.

[the remark this answers is not in the document]
Comment (scott-lewis-nsidc) [reply lib/search_solr_tools/logging/sst_logger.rb:22]: I should also note that the output for the stdout and the log file are slightly different - the stdout console doesn't put datetime stamps, whereas those ARE written to the log files.

[the remark this answers is not in the document]
Comment (scott-lewis-nsidc) [reply lib/search_solr_tools/logging/sst_logger.rb:22]: Also, is this comment in the right place? This log_environment method is just to set the environment passed into the CLI (integration, staging, etc), as the logger module would not have that information otherwise.

File: README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (juliacollins) [reply README.md:218]: Ah, my error -- I was looking at the puppet config and forgetting `share` was the destination, not the source.

File: lib/search_solr_tools/logging/sst_logger.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (juliacollins) [reply lib/search_solr_tools/logging/sst_logger.rb:23]: I suppose I could add the comment at line 13, but the gist would be the same. The name of the method gives me the wrong idea about what it's doing. I expect it to be logging information _about_ the environment. Apparently it is actually setting up the logger, including information about the environment. Why not call `new_logger(env)`? As far as I can tell `log_environment` isn't necessary.

[the remark this answers is not in the document]
Comment (juliacollins) [reply lib/search_solr_tools/logging/sst_logger.rb:23]: As far as stdout vs. the log, I don't usually expect everything added to the log to also be printed to stdout.

[the remark this answers is not in the document]
Comment (scott-lewis-nsidc) [reply lib/search_solr_tools/logging/sst_logger.rb:23]: I originally had the `log_environment` method called `log_environment=` (to emphasize you were setting the environment), but it was giving me issues in trying to filter that down from the Module to the Class.  I tried doing `set_log_environment` next, but rubocop yelled about me about having methods with `set` in the name.

`new_logger` creates the logger and passes it back, but doesn't "save" the logger as a class variable.  ``new_logger` itself is meant to be more of a private thing.

I'll see if I can find a different name for this method to make it more clear.


As for the console vs the log file, I agree they don't necessarily need to say the same things, which is why their logging levels can be separate.  In fact, for the Ping method, and other "help" methods, it doesn't write to logs at all.  But during the harvesting-related jobs, I can't think of any statement that you would potentially put on the console that you wouldn't at least want to potentially put in a log file.  Having different log levels for the console and the file allow you to configure the verbosity separately for both.

[the remark this answers is not in the document]
Comment (juliacollins) [reply lib/search_solr_tools/logging/sst_logger.rb:23]: What is the rationale for the Module vs. adding the class methods? Is this architecture something you needed to do to work around Thor? This seems a bit complex for what we need to do here.

[the remark this answers is not in the document]
Comment (juliacollins) [reply lib/search_solr_tools/logging/sst_logger.rb:23]: What I expect to see in the Thor file is a log initialization like:

logger = SSTLogger.logger(environment or whatever other initialization information is needed)
logger.info('some information')

Usually the level of stdout information is controlled by a verbosity flag. If you want to tie stdout to log content, for simplicity's sake I would just have the use the same log level.

File: lib/search_solr_tools/logging/sst_logger.rb [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    SSTLogger.logger
+  end
+
+  def log_environment(new_env)
+    SSTLogger.log_environment(new_env)
+  end
+
+  class << self
+    def logger
+      @logger ||= new_logger
```
Comment (juliacollins) [inline lib/search_solr_tools/logging/sst_logger.rb:20]: This is actually creating the logger in the SolrHarvestCLI class, right? I suggest skipping the trickiness and explicitly creating the logger as I note in a comment below. The approach here doesn't save any code because then you have to add the `log_environment` step. Make the logger setup easier to understand.

File: lib/search_solr_tools/config/environments.rb [hunk 5]
```diff
@@ -5,7 +5,7 @@
 module SearchSolrTools
   # configuration to work with solr locally, or on integration/qa/staging/prod
   module SolrEnvironments
-    YAML_ENVS = YAML.load_file(File.expand_path('environments.yaml', __dir__))
+    YAML_ENVS = YAML.load_file(File.expand_path('environments.yaml', __dir__), aliases: true)
 
     def self.[](env = :development)
       YAML_ENVS[:common].merge(YAML_ENVS[env.to_sym])
```
Comment (juliacollins) [inline lib/search_solr_tools/config/environments.rb:11]: Do you know of a reason why `common` can't be integrated into each environment in the YAML file itself (using `<<: *common`? The `dev`/`development` environment might need some refactoring, but the others could be handled in the config file itself, I think.

File: lib/search_solr_tools/logging/sst_logger.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (scott-lewis-nsidc) [reply lib/search_solr_tools/logging/sst_logger.rb:23]: The logger is intentionally a global thing - that way, every class is using the same logger without having to manually pass a logger into every initializer.  Since the "environment" in this app is not the "app environment" but rather the "target environment", I'm actually leaning towards not having environment-specific configuration at all; calling the CLI with `--environment integration` really isn't going to be any different than `--environment production` as far as the logging is concerned - the only difference is what DCS it calls and what Solr it writes to.

The "log_environment" is what is making this ugly.  If we ever have a need to do individual log setups we can re-evaluate at that point, but for now, I'll just refactor to have the logging configuration consistent.

File: lib/search_solr_tools/config/environments.yaml [hunk 6]
```diff
@@ -9,17 +9,26 @@
...
-:dev:
+:dev: &dev
   ## For the below, you'll need to instantiate your own search-solr instance, and point host to that.
   :host: dev.search-solr.USERNAME.dev.int.nsidc.org
   ## For the metadata content, either set up your own instance of dataset-catalog-services
   ## or change the URL below to point to integration
   :nsidc_dataset_metadata_url: http://dev.dcs.USERNAME.dev.int.nsidc.org:1580/api/dataset/metadata/
 
+:development:
+  <<: *dev
```
Comment (juliacollins) [inline lib/search_solr_tools/config/environments.yaml:30]: Did you add `development` for consistency with anything in particular?

File: lib/search_solr_tools/logging/sst_logger.rb (hunk 4 above)
Comment (scott-lewis-nsidc) [reply lib/search_solr_tools/logging/sst_logger.rb:20]: The logger is used in many classes; having to pass that logger around from class to class would create a lot more complication than having it be used as a global.

Comment (juliacollins) [reply lib/search_solr_tools/logging/sst_logger.rb:20]: I must be missing something, then. It looks like each class that includes the logger module would add this logger method to itself, with each creating a new `@logger` instance. I'll look again.

Comment (scott-lewis-nsidc) [reply lib/search_solr_tools/logging/sst_logger.rb:20]: Each class does include the module.  The module basically uses it's "logger" method as a pass-through to the class method; the SSTLogger module/class are never instantiated, all the methods are class-level methods.

The module's logger method just passes the call to the class method, which creates a new logger if the @logger variable hasn't been created yet, or just passes back the existing one if it has.  They all share the same logger, though.

File: lib/search_solr_tools/logging/sst_logger.rb [hunk 7]
```diff
@@ -0,0 +1,78 @@
+# frozen_string_literal: true
+
+require 'fileutils'
+require 'logging'
+require 'search_solr_tools'
```
Comment (juliacollins) [inline lib/search_solr_tools/logging/sst_logger.rb:5]: I'm apparently losing my mind. `search_solr_tools.rb` requires `search_solr_tools/logging/sst_logger` -- is this a circular situation or am I completely missing a different `search_solr_tools` that's being required here?

File: lib/search_solr_tools/logging/sst_logger.rb (hunk 4 above)
Comment (scott-lewis-nsidc) [reply lib/search_solr_tools/logging/sst_logger.rb:20]: Each class does include the Module, but the module itself is never instantiated; all the methods are class/module-level methods.

The "outer" `logger` method just allows us to call `logger.info` directly, instead of having to do `SSTLogging.logger`; that method basically just calls the "class" method called `logger`, which itself will create a new logger and assign it to `@logger`, unless the `@logger` class variable was already set (in which case, it just returns it).

So basically, the first time `logger.info` is called (by ANY class) it will instantiate the logger.  All subsequent logger calls will return the same already-instantiated logger, regardless of which class uses it.

Comment (juliacollins) [reply lib/search_solr_tools/logging/sst_logger.rb:20]: I'm not convinced of the wisdom of this approach, but I'll go with it.

File: lib/search_solr_tools/logging/sst_logger.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (juliacollins) [reply lib/search_solr_tools/logging/sst_logger.rb:23]: I'm all in favor of removing the environment.

File: lib/search_solr_tools/config/environments.yaml (hunk 6 above)
Comment (scott-lewis-nsidc) [reply lib/search_solr_tools/config/environments.yaml:30]: I could have just removed the :dev: section, but decided it didn't hurt to have both.  But in `environment.rb` it uses the default `:development` (not :dev), and I had run into some issues when trying to use the `:development` environment because of this.

I figured, having both here just means that both "dev" and "development" environment settings will behave the same way

File: lib/search_solr_tools/config/environments.rb (hunk 5 above)
Comment (scott-lewis-nsidc) [reply lib/search_solr_tools/config/environments.rb:11]: I'm not sure why the environment.yaml file was written this way with the rb file merging them.

I'll try updating the yaml file, and then the merge on line 11 wouldn't need to be there.

File: lib/search_solr_tools/logging/sst_logger.rb (hunk 7 above)
Comment (scott-lewis-nsidc) [reply lib/search_solr_tools/logging/sst_logger.rb:5]: I could probably have this just manually require the configuration itself directly.

Ruby seems to handle the circular requires just fine, but I think it would be cleaner to modify this since the logger itself doesn't need anything in the main search_solr_tools file except the configuration stuff.

Comment (juliacollins) [reply lib/search_solr_tools/logging/sst_logger.rb:5]: It's a red flag to me.

Comment (juliacollins) [reply lib/search_solr_tools/logging/sst_logger.rb:5]: What "configuration stuff" are you referring to?

Comment (juliacollins) [reply lib/search_solr_tools/logging/sst_logger.rb:5]: @scott-lewis-nsidc I see a reference to `SearchSolrTools::SolrEnvironments` in the file `bin/search_solr_tools`, but the `require` statement should be referring to `lib/search_solr_tools.rb`.

Comment (scott-lewis-nsidc) [reply lib/search_solr_tools/logging/sst_logger.rb]: I've modified the require here to specifically have the logger only load the config/environment file rather than the whole search_solr_tools module.

File: lib/search_solr_tools/config/environments.rb [hunk 8]
```diff
@@ -5,10 +5,10 @@
 module SearchSolrTools
   # configuration to work with solr locally, or on integration/qa/staging/prod
   module SolrEnvironments
-    YAML_ENVS = YAML.load_file(File.expand_path('environments.yaml', __dir__))
+    YAML_ENVS = YAML.load_file(File.expand_path('environments.yaml', __dir__), aliases: true)
 
     def self.[](env = :development)
-      YAML_ENVS[:common].merge(YAML_ENVS[env.to_sym])
+      YAML_ENVS[env.to_sym]
```
Comment (scott-lewis-nsidc) [inline lib/search_solr_tools/config/environments.rb:11]: This change, coupled with the change in environments.yaml, seems to have the same net effect but in a cleaner way.

File: lib/search_solr_tools/logging/sst_logger.rb [hunk 9]
```diff
@@ -0,0 +1,71 @@
+# frozen_string_literal: true
+
+require 'fileutils'
+require 'logging'
+
+require_relative '../config/environments'
```
Comment (scott-lewis-nsidc) [inline lib/search_solr_tools/logging/sst_logger.rb:6]: Now this only loads the environments file.

File: README.md [hunk 10]
```diff
@@ -208,6 +212,50 @@ which can be modified, or additional environments can be added by just adding a
...
+at `info` level.  These settings are configured in the `environments.yaml` config
+file.  While these settings are configured in the `common` section of that file,
+they can be overridded for specific environments by adding the configuration
```
Comment (juliacollins) [inline README.md:221]: Small typo in "overridden"

File: README.md [hunk 11]
```diff
@@ -208,6 +212,50 @@ which can be modified, or additional environments can be added by just adding a
...
+setting to the individual environment section.  **NOTE: The environment is based
```
Comment (juliacollins) [inline README.md:222]: As per our discussion, I think this should be "the `environment` setting indicates the environment used for harvesting..."
